spellcheck/platform/linux/spellcheck_linux.cpp

```diff
@@ -1,38 +1,160 @@
 // This file is part of Desktop App Toolkit,
 // a set of libraries for developing nice desktop applications.
 //
-// For license and copyright information please follow this link:
-// https://github.com/desktop-app/legal/blob/master/LEGAL
-//
+// Author: Nicholas Guriev <guriev-ns@ya.ru>, public domain, 2019
+// License: CC0, https://creativecommons.org/publicdomain/zero/1.0/legalcode
+
+#include <set>
+#include <QLocale>
+
+#include "spellcheck/spellcheck_utils.h"
+#include "spellcheck/platform/linux/linux_enchant.h"
+
 #include "spellcheck/platform/linux/spellcheck_linux.h"
 
-namespace Platform::Spellchecker {
+namespace {
+
+class EnchantSpellChecker {
+public:
+	bool isAvailable() {
+		return !_validators.empty();
+	}
+
+	auto knownLanguages() {
+		return _validators | ranges::views::transform([](const auto &validator) {
+			return QString(validator->get_lang().c_str());
+		}) | ranges::to_vector;
+	}
+
+	bool checkSpelling(const QString &word) {
+		auto w = word.toStdString();
+		return ranges::any_of(_validators, [&w](const auto &validator) {
+			return validator->check(w);
+		}) || _validators.empty();
+	}
+
+	auto findSuggestions(const QString &word) {
+		auto w = word.toStdString();
+		std::vector<QString> result;
+		for (const auto &validator : _validators) {
+			for (const auto &replacement : validator->suggest(w)) {
+				if (!replacement.empty()) {
+					result.push_back(replacement.c_str());
+				}
+				if (result.size() >= Platform::Spellchecker::kMaxSuggestions) {
+					break;
+				}
+			}
+			if (!result.empty()) {
+				break;
+			}
+		}
+		return result;
+	}
+
+	void addWord(const QString &wordToAdd) {
+		auto word = wordToAdd.toStdString();
+		auto &&first = _validators.at(0);
+		first->add(word);
+		first->add_to_session(word);
+	}
+
+	void ignoreWord(const QString &word) {
+		_validators.at(0)->add_to_session(word.toStdString());
+	}
+
+	void removeWord(const QString &word) {
+		auto w = word.toStdString();
+		for (const auto &validator : _validators) {
+			validator->remove_from_session(w);
+			validator->remove(w);
+		}
+	}
 
-bool CheckSpelling(const QString &wordToCheck) {
-	return true;
+	bool isWordInDictionary(const QString &word) {
+		auto w = word.toStdString();
+		return ranges::any_of(_validators, [&w](const auto &validator) {
+			return validator->is_added(w);
+		});
+	}
+
+	static EnchantSpellChecker *instance() {
+		static EnchantSpellChecker capsule;
+		return &capsule;
+	}
+
+private:
+	EnchantSpellChecker() {
```

Comment:
While this is less of an issue, I still recommend to use outside definitions instead of in-class definitions for the methods of such (not so small) class with a lot of (not so small) methods.